### PR TITLE
Unify `PresentMaker` and `MissingNodesBasicFormatter`, fixing a formatting bug in string literals

### LINKED ISSUE #1959: BasicFormat outputs incomplete indentation when formatting multiline string literals containing blank lines
### Description

The following code is formatted using BasicFormat (spaces are replaced with `␣` for clarity),

```swift
assertionFailure("""
␣␣First␣line

␣␣Second␣line

␣␣""")
```

It produces the following code

```swift
assertionFailure("""
␣␣First␣line
␣␣␣␣
␣␣Second␣line
␣␣␣␣
␣␣␣␣␣␣""")
```

This is an incorrectly indented literal closing quote.
This causes a compilation error `Insufficient indentation of line in multi-line string literal`.

This problem can easily occur if both the "Automatically trim trailing whitespace" and "Including whitespace-only lines" settings in Xcode are enabled.

Also, macros now automatically format the code they generate by default, which can easily cause problems if the above settings are enabled.

For example, passing the following string to the `#stringify` macro created by the Swift Macro Project template will result in a compilation error due to the same issue.

```swift
#stringify(
␣␣"""
␣␣First line

␣␣Second line

␣␣"""
)
```

^ The above macro will generate the following code, which will result in a compiation error

```swift
(
␣␣"""
␣␣First␣line
␣␣
␣␣Second␣line
␣␣
␣␣␣␣""",␣#"\#n␣␣"""\#n␣␣First␣line\#n\#n␣␣Second␣line\#n\#n␣␣""""#)
```

### Steps to Reproduce

To reproduce this problem, run the following code

```swift
let source = #"""
  assertionFailure("""
    First line

    Second line

    """)
  """#
print(Parser.parse(source: source).formatted().description)
```

The result of the execution will be as follows This code results in a compilation error "Insufficient indentation of line in multi-line string literal".

```swift
assertionFailure("""
  First line
    
  Second line
    
      """)
```

### PR DESCRIPTION
**Pull request by ahoppen**
The underlying issues of the malformatted multi-line string literal are that

1. We are setting an `anchorPoint` for empty lines. Anchor points are meant to be starting points at which the user didn’t provide any manual indentation relative to which we should format the rest of the tree. But for empty lines it’s not like the user didn’t provide any indentation. There simply wasn’t anything to indent.
2. `leadingTrivia.trimmingTrailingWhitespaceBeforeNewline` was not considering whether the token text itself contained a newline in `isBeforeNewline`.

Now, to detect if the token is followed by a newline, we should check if its text is empty, which is the case for empty string literals. Unfortunately, this is also the case for missing identifiers, which also have an empty text but whose empty text will be replace by a placeholder by `PresentMaker`. To distinguish between the two, I merged `PresentMaker` and `MissingNodesBasicFormatter` so that the `BasicFormat` also performs the text replacement to placeholders.

Fixes #1959

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

File: Tests/SwiftBasicFormatTest/BasicFormatTests.swift [hunk 1]
```diff
@@ -250,6 +250,23 @@ final class BasicFormatTest: XCTestCase {
     )
   }
 
+  func testMultilineStringLiteralWithBlackLines() {
```
Comment (kishikawakatsumi) [inline Tests/SwiftBasicFormatTest/BasicFormatTests.swift:253]:
```suggestion
  func testMultilineStringLiteralWithBlankLines() {
```

I made a typo in the test method name.

Comment (ahoppen) [reply Tests/SwiftBasicFormatTest/BasicFormatTests.swift:253]: Haha, thanks for pointing it out.

File: Tests/SwiftBasicFormatTest/BasicFormatTests.swift [hunk 2]
```diff
@@ -250,6 +250,23 @@ final class BasicFormatTest: XCTestCase {
     )
   }
 
+  func testMultilineStringLiteralWithBlankLines() {
+    assertFormatted(
+      source: #"""
+        assertionFailure("""
+          First line
```
Comment (kishikawakatsumi) [inline Tests/SwiftBasicFormatTest/BasicFormatTests.swift:257]: The blank line is gone. The code is the same as the test code above `testMultilineStringLiteral()`.

**Comment by kishikawakatsumi** [comment]
In the following test case (no spaces in the empty lines), an indent is inserted in the first blank line. This is not identical to the original string.

```swift
assertFormatted(
  source: #"""
    assertionFailure("""


      First line
      Second line
      """)
    """#,
  expected: #"""
    assertionFailure("""


      First line
      Second line
      """)
    """#
)
```

Actual Result:
```
assertionFailure("""
␣␣␣␣

  First line
  Second line
  """)
```

File: Sources/SwiftBasicFormat/BasicFormat.swift [hunk 3]
```diff
@@ -436,6 +458,19 @@ open class BasicFormat: SyntaxRewriter {
       return trailingTrivia + Trivia(pieces: nextTokenLeadingWhitespace)
     }()
 
+    /// Whether the leading trivia of the token is folloed by a newline.
```
Comment (kimdv) [inline Sources/SwiftBasicFormat/BasicFormat.swift:461]: Typo in `followed`

**Review by kimdv — APPROVED** [review]

**Review by bnbarham — APPROVED** [review]

File: Tests/SwiftBasicFormatTest/BasicFormatTests.swift (hunk 2 above)
Comment (ahoppen) [reply Tests/SwiftBasicFormatTest/BasicFormatTests.swift:257]: Oh, copying code from GitHub apparently drops empty lines 🤯 Thanks for catching it. I updated the test case

**Comment by ahoppen** [comment]
> In the following test case (no spaces in the empty lines), an indent is inserted in the first blank line. This is not identical to the original string.

Fixed

File: Sources/SwiftBasicFormat/BasicFormat.swift (hunk 3 above)
Comment (ahoppen) [reply Sources/SwiftBasicFormat/BasicFormat.swift]: Thanks. Found another two typos in the file as well.

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by kishikawakatsumi** [comment]
Running the following test case with the latest commit 3961b9c3bf3019e38c17db190de774d32b22ff61 produces incorrectly indented results.


```
assertFormatted(
  source: #"""
    assertionFailure("""

      \("""
        First Line
      """)
    """)
    """#,
  expected: ...
)
```

Actual result:

```
assertionFailure("""

      \("""
    First Line
  """)
    """)
```

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
Fixed as well. I can’t think of any more ways to break formatting of multi-line string literals but if you do find more, please let me know @kishikawakatsumi. And thanks for the thorough testing.

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test

**Pull request merged**